### PR TITLE
[C] Index enumeration values for c / cpp

### PR DESCRIPTION
**Pull request by LinuxDevon**
This is to address this issue: #689.

The previous PR: #976 is dead at the moment. There isn't any response from the original poster. This branch is intended to address the comments in the PR.

**DISCLAIMER** I haven't worked with Objective-C but was request a port to that language as well in the original PR. I did my best and based on the tests looks exactly like its counter parts C/CPP.

### CODE REVIEW

**Review by keith-hall — APPROVED** [review]

File: C++/syntax_test_c.c [hunk 1]
```diff
@@ -10,6 +10,10 @@ int main(){
 }
 
 enum Foo { kFoo, kBar };
+/* <- keyword.declaration */
+/*   ^ entity.name.enum */
+/*         ^ entity.name.enum.constants */
```
Comment (deathaxe) [inline C++/syntax_test_c.c:15]: `entity.name.enum` should be reserved for the enum data type (e.g.: `Foo`).

I'd suggest `entity.name.constant` for `kFoo` and `kBar` as this is what D and Java use.

_C# and Rust scope them `constant.other`, but this should probably be changed._

File: C++/C.sublime-syntax [hunk 2]
```diff
@@ -687,6 +687,14 @@ contexts:
       push: data-structures
     - include: expressions
 
+  data-structures-body-enum:
+    - include: preprocessor-data-structures
+    - match: '(?={{before_tag}})'
+      push: data-structures
+    - match: '{{identifier}}'
+      scope: entity.name.enum.constants
```
Comment (deathaxe) [inline C++/C.sublime-syntax:695]:
```suggestion
      scope: entity.name.constant.c
```

File: C++/C++.sublime-syntax [hunk 3]
```diff
@@ -333,6 +333,13 @@ contexts:
     - include: scope:source.c#label
     - include: expressions
 
+  statements-enum:
+    - include: preprocessor-statements
+    - include: scope:source.c#label
+    - match: '{{identifier}}'
+      scope: entity.name.enum.constants.c++
```
Comment (deathaxe) [inline C++/C++.sublime-syntax:340]:
```suggestion
      scope: entity.name.constant.c++
```

File: Objective-C/Objective-C++.sublime-syntax [hunk 4]
```diff
@@ -117,6 +117,13 @@ contexts:
     - include: scope:source.c#label
     - include: expressions
 
+  statements-enum:
+    - include: preprocessor-statements
+    - include: scope:source.c#label
+    - match: '{{identifier}}'
+      scope: entity.name.enum.constants.objc++
```
Comment (deathaxe) [inline Objective-C/Objective-C++.sublime-syntax:124]:
```suggestion
      scope: entity.name.constant.objc++
```

File: Objective-C/Objective-C.sublime-syntax [hunk 5]
```diff
@@ -783,6 +783,14 @@ contexts:
       push: data-structures
     - include: expressions
 
+  data-structures-body-enum:
+    - include: preprocessor-data-structures
+    - match: '(?={{before_tag}})'
+      push: data-structures
+    - match: '{{identifier}}'
+      scope: entity.name.enum.constants
```
Comment (deathaxe) [inline Objective-C/Objective-C.sublime-syntax:791]:
```suggestion
      scope: entity.name.constant.objc
```

**Review by deathaxe — CHANGES_REQUESTED** [review]
I am fine with this solution in general. I'd just suggest to change scope names of enum constants to get inline with D and Java.

_The general scope naming scheme is about scopes beginning with very general names such as `entity` or `constant` and each level of scope adding more detail up to the language they refer to._

Note: `entity.name.constant` should be added to `Symbol List.tmPreferences`. `Symbol Index - Include Constants.tmPreferences` can be removed then.

File: C++/syntax_test_c.c (hunk 1 above)
Comment (LinuxDevon) [reply C++/syntax_test_c.c]: Pulled in the suggested changes for the scope change. Updated the tests to reflect that as well. Thank you for the input! Let me know if there is anything else or something I missed

Comment (deathaxe) [reply C++/syntax_test_c.c]: You may probably want to pick the tmPreferences from https://github.com/LinuxDevon/Packages/pull/1 so `entity.name.constant` is properly added to symbol list and index.

**Comment by LinuxDevon** [comment]
@deathaxe I think I understood what you meant about the note with the `tmPreference` files. Double check me on that please. The Objective-C didn't match so I modified that to mirror the C/C++ one. Removed the `constant` file since these are now in the symbol file.

File: C++/syntax_test_c.c (hunk 1 above)
Comment (LinuxDevon) [reply C++/syntax_test_c.c]: Just updated the `tmPreference` files as well. I missed the comment earlier.

File: Objective-C/Symbol Index.tmPreferences [hunk 6]
```diff
@@ -2,11 +2,13 @@
 <plist version="1.0">
 <dict>
 	<key>scope</key>
-	<string>source.objc, source.objc++</string>
```
Comment (deathaxe) [inline Objective-C/Symbol Index.tmPreferences:5]: This file I also stumbled over.

It is originally meant to do some symbol translations only. Its scopes overlap with the same file in C++ package and the translation rule conflicting with the other file. Haven't looked into it any further though.

But as _C++/Symbol List.tmPreferences_ already adds `entity.name.constant` to symbol list and index for `source.objc` and `source.objc++` your change here is not needed.

Comment (LinuxDevon) [reply Objective-C/Symbol Index.tmPreferences:5]: Interesting. I didn't see the `source.objc` and `source.objc++`. So to be clear, revert this file but keep the `Symbol Index - Include Constants.tmPreference` deleted? Seeing as how the `entity.name.constant` is already added via ` C++/Symbol List.tmPreferences`

Comment (LinuxDevon) [reply Objective-C/Symbol Index.tmPreferences]: Complete

**Review by deathaxe — APPROVED** [review]

**Review by keith-hall — APPROVED** [review]

**Comment by rwols** [comment]
Someone should really just rewrite the C-family with inheritance and stuff like that 😬

**Comment by deathaxe** [comment]
I had been about starting to, several times, but than other simpler to fix issues were found again and again.

**Pull request merged**